adam/learner/cross_situational.py

```diff
@@ -0,0 +1,567 @@
+        """
+        Computes the fraction of pattern graph nodes of *pattern* which match *graph*.
+        """
+        hypothesis_pattern_common_subgraph = get_largest_matching_pattern(
+            pattern,
+            graph,
+            debug_callback=self._debug_callback,
+            graph_logger=self._graph_logger,
+            ontology=self._ontology,
+            matching_objects=True,
+        )
+        self.debug_counter += 1
+
+        leading_hypothesis_num_nodes = len(pattern)
+        num_nodes_matched = (
+            len(hypothesis_pattern_common_subgraph.copy_as_digraph().nodes)
+            if hypothesis_pattern_common_subgraph
+            else 0
+        )
+        return CrossSituationalLanguageLearner.PartialMatch(
+            hypothesis_pattern_common_subgraph,
+            num_nodes_matched=num_nodes_matched,
+            num_nodes_in_pattern=leading_hypothesis_num_nodes,
+        )
+
+    # Re-implement for cross-situational
+    def lexicon_step(self, word: str) -> None:
+        # If any conditional probability P(h^|w) exceeds a certain threshold value (h), then file
+        # (w, h^) into the
+        # lexicon
+        # From Pursuit paper: P(h|w) = (A(w,h) + Gamma) / (Sum(A_w) + N x Gamma)
+        leading_hypothesis_entry = self._leading_hypothesis_for(word)
+        assert leading_hypothesis_entry
+        (leading_hypothesis_pattern, leading_hypothesis_score) = leading_hypothesis_entry
+
+        all_hypotheses_for_word = self._words_to_hypotheses_and_scores[word]
+        sum_of_all_scores = sum(all_hypotheses_for_word.values())
+        number_of_meanings = len(all_hypotheses_for_word)
+
+        probability_of_meaning_given_word = (
+            leading_hypothesis_score + self._smoothing_parameter
+        ) / (sum_of_all_scores + number_of_meanings * self._smoothing_parameter)
+        times_word_has_been_seen = self._words_to_number_of_observations[word]
+        logging.info(
+            "Prob of meaning given word: %s, Times seen: %s",
+            probability_of_meaning_given_word,
+            times_word_has_been_seen,
+        )
+        # file (w, h^) into the lexicon
+
+        # TODO: We sometimes prematurely lexicalize words, so we use this arbitrary counter
+        #  threshold
+        if probability_of_meaning_given_word > self._lexicon_entry_threshold:
+            if times_word_has_been_seen > 5:
+                self._lexicon[word] = leading_hypothesis_pattern
+                # Remove the word from hypotheses
+                self._words_to_hypotheses_and_scores.pop(word)
+                if self._graph_logger:
+                    self._graph_logger.log_graph(
+                        leading_hypothesis_pattern, logging.INFO, "Lexicalized %s", word
+                    )
+            else:
+                logging.info("Would lexicalize, but haven't see the word often enough")
+
+    @staticmethod
+    def get_objects_from_perception(
```

Comment:
There are a couple of methods that are shared between learners. It's generally better to modularize methods like this that is shared across the code. Let's keep it this way for now though, since I'm working on a different branch changing the Pursuit class right now.